utils/check_roxie/check_roxie

```diff
@@ -1,11 +1,74 @@
-#!/bin/bash
+#!/usr/bin/env python
 
-OUTPUT=`testsocket $1 "<control:alive/>" |grep Status |grep ok -c`
+from __future__ import print_function
+import subprocess
+import re
+import sys
 
-if [ "$OUTPUT" -eq "1" ]; then
-    echo "OK"
-    exit 0
-else
-    echo "CRITICAL"
-    exit 2
-fi
+# Output usage if argument is missing
+try:
+    IP = sys.argv[ 1 ]
+except IndexError:
+    print ( "\nNeeds IP address of the target server" )
+    print ( """
+Usage:
+        check_roxie <ip-address>
+""" )
+    sys.exit ( 1 )
+
+# Nagios exit codes
+ok = 0
+warning = 1
+critical = 2
+
+cmd_roxie_status = '2>/dev/null testsocket %s "<control:state/>"' % IP
+status = subprocess.Popen ( cmd_roxie_status.strip ( ) , stdout=subprocess.PIPE , shell=True )
+status_result = status.communicate ( )[ 0 ]
+roxie_status = re.findall ( '\<Status\>(.+?)\</Status\>+?' , status_result )
+my_hash = re.findall ( "hash='(.*?)'" , status_result )
+
+# Check if roxie process is up and running
+if roxie_status and roxie_status[ 0 ] == 'ok':
+    # Get state hashes from roxie cluster
+    hashes = [ ]
+    try:
+        cmd_cluster_state = '2>/dev/null testsocket %s -lock "<control:state/>"' % IP
+        cluster_state = subprocess.Popen ( cmd_cluster_state.strip ( ) , stdout=subprocess.PIPE , shell=True )
+        cluster_state_result = cluster_state.communicate ( )[ 0 ]
+        for cluster_result in cluster_state_result.split():
+            if "hash" in cluster_result:
+                state_hash = re.findall ( "hash=\"(.*?)\"" , cluster_result )
+                hashes.append ( state_hash[ 0 ] )
+    except:
+        pass
+
+    total_hashes = len ( hashes )
+    same_hash = [ ]
+    different_hash = [ ]
+
+    # Compare hash state between self and cluster
+    while total_hashes > 0:
+        for i in hashes:
+            if my_hash[ 0 ] == i:
+                same_hash.append ( i )
+                total_hashes -= 1
+            elif my_hash[ 0 ] != i:
+                different_hash.append ( i )
+                total_hashes -= 1
+        if len ( same_hash ) < len ( different_hash ) and my_hash[ 0 ] in same_hash:
+            print ( "Roxie process is running but there is a state mismatch " ,
+                    my_hash[ 0 ].replace ( '\n' , '' ) )
+            sys.exit ( warning )
+        elif len ( same_hash ) < len ( different_hash ) and my_hash[ 0 ] not in same_hash:
+            print ( "Roxie process is running but there is a state mismatch " ,
+                    my_hash[ 0 ].replace ( '\n' , '' ) )
+            sys.exit ( warning )
+        elif len ( same_hash ) > len ( different_hash ) and my_hash[ 0 ] in same_hash:
```

Comment:
If I understanding the logic well this isn't a proper condition. If we have 3 same hashes and 1 different then can we say "state hash matches across roxie cluster"?

If not then I think 
```
    elif  len( different_hash ) == 0  and my_hash[ 0 ] in same_hash:
````
of what I suggested above:
```
   elif  numOfDifferentHashes == 0  and myHashMatchWithOthers:
```
would be the proper condition.
Or going a step further perhaps
```
   elif  numOfDifferentHashes == 0
```
would be enough. I think.

Reply:
@AttilaVamos 
The objective is to reduce the number of false positives sent by Nagios.  

if not myHashMatchWithOthers: 
elif numOfSameHashes < numOfDifferentHashes and myHashMatchWithOthers:  

The 2 conditions shown above will trigger a 'warning' Nagios alert but only for the servers that fulfill any of those conditions.

while the one shown below will not  trigger an email alert.

elif numOfSameHashes > numOfDifferentHashes and myHashMatchWithOthers: